File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/entities/ModelProvidingBlockEntity.java
package generations.gg.generations.core.generationscore.common.world.level.block.entities;

import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockAnimatedObjectInstance;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockObjectInstance;
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockAnimatedObjectInstance;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockObjectInstance;
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockAnimatedObjectInstance;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockObjectInstance;
import generations.gg.generations.core.generationscore.common.world.level.block.generic.GenericModelBlock;
import gg.generations.rarecandy.renderer.rendering.ObjectInstance;
import net.minecraft.core.BlockPos;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.AABB;
import org.joml.Matrix4f;

public abstract class ModelProvidingBlockEntity extends SimpleBlockEntity implements ModelContextProviders.ModelProvider, ModelContextProviders.VariantProvider {
    public ObjectInstance[] objectInstance;
    private AABB boundingBox;

    public ModelProvidingBlockEntity(BlockEntityType<? extends ModelProvidingBlockEntity> type, BlockPos pos, BlockState state) {
        super(type, pos, state);
    }

    public ResourceLocation getModel() {
        return (getBlockState().getBlock() instanceof GenericModelBlock<?> block ? block.getModel() : GenerationsBlockEntityModels.DEFAULT);
    }

    public AABB getRenderBoundingBox() {
        if(boundingBox == null) boundingBox = getBlockState().getBlock() instanceof GenericModelBlock<?> block ? block.computeRenderBoundingBox(getLevel(), getBlockPos(), getBlockState()) : defaultAABB(getBlockPos());
        return boundingBox;
    }

    @Override
    public String getVariant() {
        return getBlockState().getBlock() instanceof ModelContextProviders.VariantProvider provider ? provider.getVariant() : null;
    }

    public static AABB defaultAABB(BlockPos pos) {
        return new AABB(pos, pos.offset(1,1,1));
    }

    public ObjectInstance generateInstance() {
        return isAnimated() ? new BlockAnimatedObjectInstance(new Matrix4f(), new Matrix4f(), null): new BlockObjectInstance(new Matrix4f(), new Matrix4f(), null);
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/entities/ModelProvidingBlockEntity.kt
package generations.gg.generations.core.generationscore.common.world.level.block.entities

import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders.VariantProvider
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockAnimatedObjectInstance
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockObjectInstance
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.instanceOrNull
import generations.gg.generations.core.generationscore.common.world.level.block.generic.GenericModelBlock
import gg.generations.rarecandy.renderer.rendering.ObjectInstance
import net.minecraft.core.BlockPos
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.Vec3
import org.joml.Matrix4f

abstract class ModelProvidingBlockEntity(
    type: BlockEntityType<out ModelProvidingBlockEntity>,
    pos: BlockPos,
    state: BlockState
) : SimpleBlockEntity(type, pos, state), ModelContextProviders.ModelProvider, VariantProvider {
    var objectInstance: Array<ObjectInstance?>? = null
    private var boundingBox: AABB? = null

    override fun getModel(): ResourceLocation = blockState.block.instanceOrNull<GenericModelBlock<*>>()?.model ?: GenerationsBlockEntityModels.DEFAULT

    val renderBoundingBox: AABB
        get() {
            if (boundingBox == null) boundingBox = blockState.block.instanceOrNull<GenericModelBlock<*>>()?.computeRenderBoundingBox(
                    getLevel()!!,
                    blockPos,
                    blockState
                ) ?: defaultAABB(blockPos) else boundingBox

            return boundingBox as AABB
        }

    override fun getVariant(): String? = blockState.block.instanceOrNull<VariantProvider>()?.variant

    fun generateInstance(): ObjectInstance {
        return if (isAnimated) BlockAnimatedObjectInstance(Matrix4f(), Matrix4f(), null) else BlockObjectInstance(
            Matrix4f(), Matrix4f(), null
        )
    }

    companion object {
        fun defaultAABB(pos: BlockPos): AABB {
            return AABB(Vec3.atLowerCornerOf(pos), Vec3.atLowerCornerOf(pos.offset(1, 1, 1)))
        }
    }
}
